### ISSUE TITLE
Better mono repo support

### ISSUE THREAD

**Issue by Izhaki** [martpie/next-transpile-modules#57]
We are using this plugin on a monorepo, so `packages/a` can import from `packages/b` without having to build `b` every time a change is made.

One of the packages is called `core`. 

# Issues we had

## Package name

When providing the package name:

```
    [withTM, { transpileModules: ['core' }],
```

That would also match `node_modules\core-js` and we'd get all sorts of (very odd) errors (Max Stack etc.).

## Full path

When providing full path, using:

```
const getPackageSourcePath = packageName => path.resolve(__dirname, `packages/${packageName}/src`);
```

This works on a MAC, but not on Windows. The matchers withTM generates:

```
include:
 [ /C:\projects\tools\packages\core\src(?!.*node_modules)/ ]
```

## Solution

After 4 hours of trail and error, we finally got it working using:

```
    [withTM, { transpileModules: ['packages/core' }],
```

---

Not really sure how to address this. This package seems to be aimed at `node_modules` rather than monorepos. But thought to report anyhow.

**Comment by martpie** [comment]
> This package seems to be aimed at node_modules rather than monorepos.

This package is perfectly suited for monorepos (though the `core` conflict may be a bug of `next-transpile-modules`, I need to investigate). I encourage you to have a look at the FAQ (I hope you see the warning on the issue template), there may be answers for you there (basically, npm local install and yarn workspaces works nice, but the use of lerna is discouraged).

**Comment by martpie** [comment]
I'll add one thing: even though you have a monorepo, you _should_ (imho) still install your local modules in `node_modules`, as you may not deploy all applications on production.

**Comment by martpie** [comment]
Bonus:

- https://github.com/zeit/next.js/issues/9295
- https://github.com/martpie/monorepo-typescript-next-the-sane-way

**Comment by Izhaki** [comment]
@martpie I admit I have jumped to conclusions when I said this package "seems to be aimed at node_modules rather than monorepos". Sorry about this.

Let me just lay it out, so to help myself, yourself, and others to understand the greater context.

# Monorepo tutorials leave you wanting

First, there's lerna, which is a huge confound- at least if you use yarn workspaces, lerna is reserved for the publishing stage only.

But the main problem is that people expect to setup yarn workspaces and everything to work, but many things do not work.

The reason for this is that yarn workspaces will only work (as most people expect) **after you build the various packages**. Even this can be tricky because you need to build both dev (with source-maps) and production versions. And still somethings may not work as expected.

# The missing link

The thing that most tutorials fail to mention, and most people don't have in their mental model is that **you want your various packages linked before you build them**.

Whether you want to click a script in another package in your IDE, or types to work, you want your packages linked.

This means and interlink between `src` folders (pre-build), not `dist` folder (post-build).

As I said, you could, in theory, set up a background build for *all* packages - but that's just bonkers and unnecessary and terribly hard to get right (all taken into account).

This missing link is called `aliases` - a way to tell the module resolver that when you see `X` you should go to `Y/X`.

There's some example of it in this [minimal typescript repo](https://github.com/Izhaki/mono.ts), and in [this (hybrid, but mainly non-typescript) repo](https://github.com/Izhaki/regraph). Note that in the latter, although there is a build for the various packages - it isn't used in practice. Instead, the `/docs` package just import things from other packages and bundles them in. Such will be the case for months ahead (I must say that conceptually **aliases should not be used in production builds**, at least not between packages, but `/docs` is a product not a package so this is OK there).

# next-transpile-modules

As far as this package is concerned, the intent is very clear. You want to tell next:

> I want stuff in X or Y folder to be build as well.

The question is:

> How do we express these folders (or modules)

The issue I have raised is split into two practical concerns:

- Why `core` also catches `core-js` (exact folder match needed maybe)?
- Why the full path doesn't work on Windows? (it worked for months on Mac until a new developer joined using Windows).

(I don't have a Window machine at my disposal, but I'm dying to know why the regex `/C:\projects\tools\packages\core\src(?!.*node_modules)/` didn't work).

# A more conceptual note

We had to log and inspect the generated webpack config, and look on the code of this library to figure out where the problem is and what sort of transformation `next-transpile-modules` applies. 
It took half a day to solve the issue we had (the error was that the build on windows complained about unrecognised character in a jsx file - a file in `core` imported from a next package).

One way to look at it is that `next-transpile-modules` has some "hidden logic" so to speak - it works great when it does, but hard to troubleshoot when it doesn't. Here's a question:

> Why do any transformations at all? Why not just take as parameter an array of include matchers that are given to the webpack rule as is?

I think the answer is obvious, but I'm still asking it to stimulate critical thinking.

**Comment by martpie** [comment]
Ohw that escalated quickly. Let's refocus a little bit 😄 (I have good intentions, I swear ✌️).

I will be investing the `core` matching conflict first. I also have a bunch of other issues to tackle, including one opened by the lead maintainer of Next for the future native CSS support of Next.js.

I will then work on the other important things for the v3.0 of this plugin (https://github.com/martpie/next-transpile-modules/milestone/1), then I'll get back to what you wrote above. I know some of the aspects of this plugin can be confusing, but they are a trade-off of a lot of different things, + legacy things from the initial code this fork is based on. But I won't feel confident until I have end-to-end tests in place.

If I can get rid of Regex, trust me, I will do it, they may things awfully complicated. I will try to do these this week, but cannot guarantee (it is not the only project I maintain, and there's work too).

I hope this if fine for you :) if not, please fork or submit a fix.

**Comment by martpie** [comment]
Hi again @Izhaki,

I can confirm there is a conflict with `core` / `core-js`, I will fix this case and add a test for this scenario

**Comment by martpie** [comment]
Hello 👋 

The `core-js` conflict will be resolved in 3.0 (the PR is over there: #58).

Regarding the full path issue, it is likely to not even be working anymore on macOS (funnily enough, partly because of the fix for `core-js` conflict). I highly recommend you to link your modules, use Yarn workspaces, on `npm i` them :)
